main.go

```diff
@@ -167,19 +161,19 @@ func filterBucketContent(elems s3.ListObjectsOutput, csvSuffix, processedFlagSuf
 	return unprocessed, processed, orphanFlags
 }
 
-func parallelApply(list s3.ListObjectsOutput, fn func(o types.Object)) {
+func parallelApply(ctx context.Context, list s3.ListObjectsOutput, fn func(o types.Object) error) {
 	//Limit the number of parallel routines doing the processing.
-	semaphore := make(chan struct{}, opts.MaxRoutines)
+	g, ctx := errgroup.WithContext(ctx)
+	g.SetLimit(opts.MaxRoutines)
+
 	var wg sync.WaitGroup
 	wg.Add(len(list.Contents))
 	for _, item := range list.Contents {
 		o := item
-		go func() {
-			semaphore <- struct{}{}        // Lock channel
-			defer func() { <-semaphore }() // Release channel
+		g.Go(func() error {
 			defer wg.Done()
-			fn(o)
-		}()
+			return fn(o)
+		})
 	}
 	wg.Wait()
```

Comment:
If you switch to an err group, no need for a waitgroup 
```
	for _, item := range list.Contents {
		o := item
		g.Go(func() error {
			fn(o)
		})
	}
	g.Wait()
```

Reply:
Indeed, should be simpler now